src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -335,20 +343,49 @@ public byte[] Http2AltSvcOriginUri
         private ValueTask<(HttpConnectionBase? connection, bool isNewConnection, HttpResponseMessage? failureResponse)>
             GetConnectionAsync(HttpRequestMessage request, bool async, CancellationToken cancellationToken)
         {
+            // Do not even attempt at getting/creating a connection if it's already obvious we cannot provided the one requested.
+            if (request.VersionPolicy != HttpVersionPolicy.RequestVersionOrLower)
+            {
+                if (request.Version.Major >= 3 && !_http3Enabled)
+                {
+                    throw new NotSupportedException("ToDo: Message=Requesting HTTP version {0} with version policy {1} while HTTP/{2} is not enabled. See '{3}' AppContext switch.");
+                }
+                if (request.Version.Major >= 2 && !_http2Enabled)
+                {
+                    throw new NotSupportedException("ToDo: Message=Requesting HTTP version {0} with version policy {1} while HTTP/{2} is not enabled. See '{3}' AppContext switch.");
+                }
+            }
+
             if (_http3Enabled && request.Version.Major >= 3)
             {
+                // ToDo: this depends on _assumePrenegotiatedHttp3ForTesting, why is the H3 connection under this condition?
```

Comment:
"Prenegotiated HTTP/3" is to connect directly to a HTTP/3 server without any version negotiation. We also support the more standard path, an ALT-SVC upgrade.

Each path sets the `_http3Authority` field.

Reply:
Does **Alt-Svc** upgrade handling have an impact on this issue? Does this mean that even when I'm getting HTTP/1.1 connection I might be able to upgrade higher?
I see in [code](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L801) that we return the original response when we handle the **Alt-Svc**.
However, if I'm throwing when I cannot get H3 connection (for `Exact` and `OrUpper`), I will never allow the request to be sent and never get the **Alt-Svc** header.
The caller would have to either allow downgrade first or sent HTTP/1.1 request. Is this expected and/or good enough behavior for this issue?

Reply:
Following up on discussion--

The version policy in this case is not isolated to one request but rather the entire connection pool. If user sets OrUpper and H1, the first request will go over H1/H2 (based on ALPN), the server will send an Alt-Svc header, and the second request will see `_http3Authority` is non-null and go over H3.